### ISSUE TITLE
Get rid of the node.js domains

### ISSUE THREAD

**Issue by pofider** [jsreport/jsreport#151]
node.js domains are about to be deprecated. jsreport should get rid of it.
Some comments about this topic are here:
https://github.com/jsreport/jsreport-scripts/commit/054dcfc67e0121829382961dacb8150d4b027f39#commitcomment-14692075

**Comment by bjrmatos** [comment]
a big :+1:

**Comment by pofider** [comment]
Main purpose of domains usage in jsreport is to supply Thread Local Storage. TLS is used to store the current express request object and reach it from any point of code. 

The solution seems to be to explicitly forward the current request object through whole call stack. This is doable, but problem occurs with event listeners and other hooks.

1.
[node-simple-odata-server](https://github.com/pofider/node-simple-odata-server) currently doesn't forward the original request object to the hooks `beforeQuery`, `beforeUpdate`, `beforeRemove` and `beforeInsert`. Luckily I maintain the library. I need to change it's API to forward the request object to the hooks. This will be breaking, but quite easy to do. 

2.
jsreport document store API doesn't forward request information to the listeners. 
```js
//waiting for someone to search for template
jsreport.documentStore.collection('templates').beforeFindListeners.add('foo', function(query) {
   //I have no idea what is the current request
})
....
//trying to search for a template
jsreport.documentStore.collection('templates').find({ name: 'foo' }).then(...)
```

The information about current request is needed in the listener. For example authorization extension needs to be able to add listener which reject the operation based on the current user.

a) The solution could be to add extra parameter to all calls
```js
jsreport.documentStore.collection('templates').find(context, { name: 'foo' })
```
as well to all listeners
```js
jsreport.documentStore.collection('templates').beforeFindListeners.add('foo', function(context, query) {
    //now I can reach the user
   context.req.user
})
```

b) Or another solution to create a wrapper around document store which will include request to all calls
```js
var session = jsreport.documentStore.createSession(request)
session.collection('templates').find({ name: 'foo'})
```
and in the listener reach the request through `this`

<hr>
Hmmm. Luckily we maintain all the problematic code. However I am not sure if these changes are the right way to go. Any suggestions? Is the problem clear?

**Comment by bjrmatos** [comment]
> a) The solution could be to add extra parameter to all calls
> 
> jsreport.documentStore.collection('templates').find(context, { name: 'foo' })
> as well to all listeners
> 
> jsreport.documentStore.collection('templates').beforeFindListeners.add('foo', function(context, query) {
>     //now I can reach the user
>    context.req.user
> })

i like this option, if some listeners need to use information in the request (and the request), just pass it.

in the end i think the code will feel more natural (because is explicit) and the available parameters for listeners are clear (without knowing about `this` or `domains`)

i don't like option b because `this` in JS is dynamic/tricky, so better don't depend on it for callbacks.

**Comment by bjrmatos** [comment]
> jsreport.documentStore.collection('templates').find(context, { name: 'foo' })

another option for `documentStore` methods could be:

```js
jsreport.documentStore.collection('templates').context({...}).find({ name: 'foo' })
```

**Comment by pofider** [comment]
Changes released in 0.12.0
